### PR TITLE
Move server/config to core

### PR DESCRIPTION
**Pull request by igor-dv**
## What I did

- Moved the server/config to core
- Deleted few deprecated things (default webpack configs, addons from app/react, head.html fallback)

TBD: RN

### CODE REVIEW

**Comment by Hypnosphi** [comment]
When working on RN, keep in mind that most of the differences between RN and react configs are only because someone forgot to apply react changes to RN

**Comment by igor-dv** [comment]
@Hypnosphi, I feel a bit unconfident to make changes in the RN part. There are still a lot of things that were not moved before (like middleware). Maybe you or @Gongreg will help with this? If not I will force myself to configure RN environment on my machine 🤦‍♂️

**Comment by Hypnosphi** [comment]
> configure RN environment on my machine

With `crna-kitchen-sink`, there's actually nothing to configure except for intalling Expo app on your phone

**Review by danielduan — DISMISSED** [review]
Please update the Webpack imports for the docs:
https://storybook.js.org/configurations/custom-webpack-config/

**Comment by igor-dv** [comment]
> Please update the Webpack imports for the docs:

@Hypnosphi already changed it before, it just wasn't released - 
https://github.com/storybooks/storybook/blob/master/docs/src/pages/configurations/custom-webpack-config/index.md#full-control-mode--default

**Comment by Gongreg** [comment]
Hey @igor-dv, I haven't touched webpack for quite a while, so I am not really confident on changing it too.

**Comment by igor-dv** [comment]
Hey @Gongreg, all is good. I will be back to it soon. 

P.S. I think I won't be able to touch it until the next week, so if one decides to review/merge it I will open a new branch for RN.

**Comment by danielduan** [comment]
Everything under the RN `/server` folder is the same as the react one. All the react-native stuff is in `/bin` and `/preview`

These webpack and babel stuff are okay to change.
https://github.com/storybooks/storybook/tree/master/app/react-native/src/server

**Comment by Hypnosphi** [comment]
So is it ready for review @igor-dv ?

**Comment by igor-dv** [comment]
I think you can review. I will need to perform the changes to Mithril (and RN) anyway. But will have time the next week.

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
Absolutely fantastic @igor-dv 🎉

File: app/angular/src/server/config.js [hunk 1]
```diff
@@ -1,97 +1,37 @@
-/* eslint-disable global-require, import/no-dynamic-require */
-import fs from 'fs';
-import path from 'path';
+import { configLoaderCreator } from '@storybook/core/server';
 import { logger } from '@storybook/node-logger';
-import { createDefaultWebpackConfig } from '@storybook/core/server';
-import loadBabelConfig from './babel_config';
```
Comment (Hypnosphi) [inline app/angular/src/server/config.js]: You didn't remove this file, why?

File: app/vue/addons.js [hunk 2]
```diff
@@ -1,8 +0,0 @@
-// import deprecate from 'util-deprecate';
-// import '@storybook/addon-actions/register';
-// import '@storybook/addon-links/register';
```
Comment (Hypnosphi) [inline app/vue/addons.js]: Please remove from `package.json`s all the dependencies that are no more used

File: app/angular/src/server/config.js (hunk 1 above)
Comment (Hypnosphi) [reply app/angular/src/server/config.js]: Nevermind, GitHub just treated it as moved to core

File: app/angular/src/server/config.js [hunk 3]
```diff
@@ -1,97 +1,37 @@
...
-    resolve: {
-      ...config.resolve,
-      ...customConfig.resolve,
-      alias: {
-        ...config.alias,
-        ...(customConfig.resolve && customConfig.resolve.alias),
-      },
-    },
-  };
-}
```
Comment (Hypnosphi) [inline app/angular/src/server/config.js]: I just realized that we don't apply angular cli config to extend mode. Are there any good reasons for that?

**Review by Hypnosphi — CHANGES_REQUESTED** [review]

File: app/vue/addons.js (hunk 2 above)
Comment (igor-dv) [reply app/vue/addons.js]: Removed `json5` and `find-cache-dir`. Didn't find others

File: app/angular/src/server/config.js (hunk 3 above)
Comment (igor-dv) [reply app/angular/src/server/config.js]: I don't remember already 😓

File: app/angular/src/server/config.js [hunk 4]
```diff
@@ -1,97 +1,37 @@
...
-    return customConfig(
-      applyAngularCliWebpackConfig(config, cliWebpackConfigOptions),
-      configType,
-      defaultConfig
-    );
-  }
-  logger.info('=> Loading custom webpack config (extending mode).');
+    return applyAngularCliWebpackConfig(config, cliWebpackConfigOptions);
+  },
+  wrapBasicConfig: config => applyAngularCliWebpackConfig(config, cliWebpackConfigOptions),
```
[the remark this answers is not in the document]
Comment (igor-dv) [reply app/angular/src/server/config.js]: Yeah. Just wanted it to happen in the same lifecycle as before. Any suggestions?

File: addons/storyshots/package.json [hunk 5]
```diff
@@ -37,6 +37,7 @@
   },
   "peerDependencies": {
     "@storybook/addons": "4.0.0-alpha.1",
+    "@storybook/core": "4.0.0-alpha.1",
```
Comment (Hypnosphi) [inline addons/storyshots/package.json]: We shouldn't add it as peerDependency unless users are supposed to install it on their side

File: app/angular/src/server/config.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hypnosphi) [reply app/angular/src/server/config.js]: detect cli before everything else?

[the remark this answers is not in the document]
Comment (igor-dv) [reply app/angular/src/server/config.js]: You convinced me 😉

**Review by danielduan — APPROVED** [review]

File: app/vue/addons.js (hunk 2 above)
Comment (Hypnosphi) [reply app/vue/addons.js]: `@storybook/addon-actions` & `@storybook/addon-links`

Comment (Hypnosphi) [reply app/vue/addons.js]: Probably `util-deprecate` as well (if there are usages, they can be removed)

Comment (igor-dv) [reply app/vue/addons.js]: > `@storybook/addon-actions` & `@storybook/addon-links`

Were only in app/react -> done

> util-deprecate

done

File: app/angular/src/server/config.js (hunk 3 above)
Comment (igor-dv) [reply app/angular/src/server/config.js]: @danielduan , what do you mean?

**Comment by ndelangen** [comment]
Needs a lint fix:
```
/tmp/storybook/app/react/src/demo/Welcome.js
  4:1  error  '@storybook/addon-links' should be listed in the project's dependencies. Run 'npm i -S @storybook/addon-links' to add it  import/no-extraneous-dependencies
```

**Review by Hypnosphi — APPROVED** [review]

**Comment by ndelangen** [comment]
screen shot 2018-04-03 at 13 55 49

👍 👍 💯

**Comment by igor-dv** [comment]
Deleting code is much more satisfying than writing code =)

**Comment by ndelangen** [comment]
Removing code is often a sign of removing complexity..

File: lib/core/src/server/config.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  };
+}
+
+function informAboutCustomConfig(defaultConfigName) {
+  if (!defaultConfigName) {
+    logger.info('=> Using default webpack setup.');
+    return;
+  }
+
+  logger.info(`=> Using default webpack setup based on "${defaultConfigName}".`);
```
Comment (Hypnosphi) [inline lib/core/src/server/config.js]: Actually, default config is always the same, and it's based on create-react-app. So when we say that it's based e.g. on polymer-cli, it's not true

**Pull request merged**